src/smt/witness_form.h

```diff
@@ -0,0 +1,94 @@
+/*********************                                                        */
+/*! \file witness_form.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The module for managing witness form conversion in proofs
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__SMT__WITNESS_FORM_H
+#define CVC4__SMT__WITNESS_FORM_H
+
+#include <unordered_set>
+
+#include "expr/node_manager.h"
+#include "expr/proof.h"
+#include "expr/proof_generator.h"
+#include "expr/term_conversion_proof_generator.h"
+
+namespace CVC4 {
+namespace smt {
+
+/**
+ * The witness form proof generator, which acts as a wrapper around a
+ * TConvProofGenerator for adding rewrite steps for witness introduction.
+ *
+ * The proof steps managed by this class are stored in a context-independent
+ * manager, which matches how witness forms are managed in SkolemManager.
+ */
+class WitnessFormGenerator : public ProofGenerator
+{
+ public:
+  WitnessFormGenerator(ProofNodeManager* pnm);
+  ~WitnessFormGenerator() {}
+  /**
+   * Get proof for, which expects an equality of the form t = toWitness(t).
+   * This returns a proof based on the term conversion proof generator utility.
+   * This proof may contain open assumptions of the form:
+   *   k = toWitness(k)
+   * Each of these assumptions are included in the set getWitnessFormEqs()
+   * below after returning this call.
+   */
+  std::shared_ptr<ProofNode> getProofFor(Node eq) override;
+  /** Identify */
+  std::string identify() const override;
+  /**
+   * Does the rewrite require witness form conversion?
+   * When calling this method, it should be the case that:
+   *   Rewriter::rewrite(toWitness(t)) == Rewriter::rewrite(toWitness(s))
+   * This method returns false if:
+   *   Rewriter::rewrite(t) == Rewriter::rewrite(s)
+   * which means that the proof of the above fact does not need to do
+   * witness form conversion to prove conclusions of MACRO_SR_PRED_TRANSFORM.
```

Comment:
Clarify before the end of the documentation that this is intended to be used for conclusions of this rule.